src/main/java/duke/component/DukeDatabase.java

```diff
@@ -0,0 +1,144 @@
+
+        // Creates the database file and the corresponding directories if they have not existed yet.
+        try {
+            tasksFile = new File(databaseDirectory);
+            tasksFile.getParentFile().mkdirs();
+            tasksFile.createNewFile();
+        } catch (IOException e) {
+            System.out.printf("Fatal Database Error during initialization! Please restart the bot!");
+        }
+
+        return this;
+    }
+
+    /**
+     * Returns all the tasks stored in the database.
+     *
+     * @return a list of all the tasks stored in the database.
+     */
+    public TaskList getAllTasks() {
+        TaskList tasksList = new TaskList();
+
+        try {
+            File f = new File(databaseDirectory);
+            Scanner sc = new Scanner(f);
+
+            // Read the database file.
+            while (sc.hasNextLine()) {
+                try {
+                    // Create the corresponding task object from the data in the database
+                    // and add into the taskList.
+                    tasksList.addTask(createTask(sc.nextLine()));
+                } catch (DukeException | NullPointerException | NumberFormatException e) {
+                    System.out.println("Database has corrupted data!");
+                }
+            }
+        } catch (FileNotFoundException e) {
+            System.out.println("Cannot find database file!");
+        }
+
+        return tasksList;
+    }
+
+    /**
+     * Creates the corresponding task object from a line of data in the database.
+     *
+     * @param dataInput a line of data in the database.
+     * @return the corresponding task object.
+     * @throws DukeException If the given data is corrupted.
+     */
+    private Task createTask(String dataInput) throws DukeException {
+        // Get the type of the Task.
+        String[] arr = dataInput.split("\\|");
+        String type = arr[0].trim();
+
+        // Create the corresponding task object.
+        Task task;
+        if ("T".equals(type)) {
+            task = new ToDo(arr[2].trim());
+        } else if ("D".equals(type)) {
+            task = new Deadline(arr[2].trim(), arr[3].trim());
+        } else if ("E".equals(type)) {
+            task = new Event(arr[2].trim(), arr[3].trim());
+        } else {
+            throw new DukeException();
+        }
+
+        // Check and update the status of the task accordingly.
+        int status = Integer.parseInt(arr[1].trim());
+        if (status == 1) {
+            task.markAsDone();
+        } else if (status != 0) {
+            throw new DukeException();
+        }
+
+        return task;
+    }
+
+    /**
+     * Updates the database file (used at the end of the program).
+     *
+     * @param tasks a list of tasks as updated at the point when the program ends.
+     */
+    public void update(TaskList tasks) {
+        // Extract all the data summaries of the tasks and append them into a single string.
```

Comment:
A few things I noticed here:

1. If program terminates (assume an unhandled exception) without calling `exit()` in ExitCommand.java, my changes would not be saved. An easy fix would be to add the following to your Duke.java:run() try-catch block:

`
...
finally {
    database.update(taskList);
}

2. Should my list size grow to the thousands (though unlikely), the list would take longer to write to file.

For 2. you could write an `append(int index, String data)` method that appends the new task into the file. Since each line denotes one task, and they are in running order, it would be easy for us to "insert" a new deadline. This reduces our write operation to a single line rather than n lines.

Reply:
Thanks for the suggestion! Regarding point 2, I understand where you're coming from, however, it can become an issue when we want to update and delete a certain task in the list. If we want to only delete or update a task in the list and then we will need to go through the entire data file in order to do so as we could not just operate on a certain line in the data file without going through the rest of the lines in it, as pointed in some StackOverflow posts. Thus, I choose to only write the file at the end of the program to reduce the overall number of write operations needed.

Anyway, thanks for the suggestion on point 1 and I'll look into it if I have the time!